lib/filewatch/tail_base.rb

```diff
@@ -52,183 +30,165 @@ def initialize(opts={})
       if !@opts.include?(:sincedb_path)
         raise NoSinceDBPathGiven.new("No HOME or SINCEDB_PATH set in environment. I need one of these set so I can keep track of the files I am following.")
       end
-      @watch.exclude(@opts[:exclude])
-      @watch.close_older = @opts[:close_older]
-      @watch.ignore_older = @opts[:ignore_older]
-      @watch.delimiter = @opts[:delimiter]
-      @watch.max_open_files = @opts[:max_open_files]
+      @watch = build_watch(@opts, @logger)
       @delimiter_byte_size = @opts[:delimiter].bytesize
+    end
 
-      _sincedb_open
-    end # def initialize
+    def build_watch(opts, logger)
+      discoverer = Discover.new(opts, logger)
+      @sincedb = SinceDb.new(opts, logger)
+      discoverer.add_converter(SinceDbConverter.new(@sincedb, logger))
+      watch = Watch.new(opts).add_discoverer(discoverer)
+      watch.max_open_files = opts[:max_open_files]
+      watch
+    end
 
-    public
     def logger=(logger)
       @logger = logger
       @watch.logger = logger
+      @sincedb.logger = logger
     end # def logger=
 
-    public
     def tail(path)
       @watch.watch(path)
     end # def tail
 
-    public
-    def sincedb_record_uid(path, stat)
-      # retain this call because its part of the public API
-      @watch.inode(path, stat)
-    end # def sincedb_record_uid
+    def sincedb_write(reason=nil)
+      @sincedb.write(reason)
+    end
+
+    # quit is a sort-of finalizer,
+    # it should be called for clean up
+    # before the instance is disposed of.
+    def quit
+      @watch.quit # <-- should close all the files
+    end # def quit
+
+    # close_file(path) is to be used by external code
+    # when it knows that it is completely done with a file.
+    # Other files or folders may still be being watched.
+    # Caution, once unwatched, a file can't be watched again
+    # unless a new instance of this class begins watching again.
+    # The sysadmin should rename, move or delete the file.
+    def close_file(path)
+      @watch.unwatch(path)
+      sincedb_write
+    end
 
     private
 
     def _open_file(watched_file, event)
```

Comment:
can we better document the responsibilities of this method_ it seems the actual opening happens in line 85, but otherwise there's also adding to the sincedb and watching